### ISSUE TITLE
Consistent Model of empty Readers and DataSets

### ISSUE THREAD

**Issue by mpkocher** [PacificBiosciences/pbcore#36]
I've talked with @dalexander and he suggested opening up an issue to define a general model for handling empty files, both at the file level (i.e., Reader) and DataSet level. 

Misc comments:
- RS-era: Empty cmp.h5 files were handled with special logic, specifically in the gather task.
- RS-era: Similar cmp.h5 "is empty" logic was in different tools/libs. 
- RS-era: Filtering was a core task and as such, the downstream reports had special logic to generate empty reports
- the current mapping_stats works fine with an empty AlignmentSet (`<ExternalResources /> <DataSets />`. It might work with an empty bam and pbi file.
- Some empty files make don't make sense, for example, the ReferenceSet must have a non-empty fasta file. This is constrained via `fasta-to-reference` but not fundamentally at the  DataSet XSD level. 
- Clarify model for Importing empty DataSet(s) at the service level from "successful" jobs
- Empty AlignmentSets in the context of Multi-job analysis might create headaches. I would have to look into the R code to see how empty files are handled in the RS-era.

**Comment by dalexander** [comment]
*TL;DR*: don't touch the behavior for basic file types; negotiate what you want, for datasets.

For basic biofx file types, I consider a file to consist of "headers/metadata" (which can include, for example, the movie table in the cmp.h5 format, or the text headers in a GFF and "records".  

A file can be empty in two possible ways: 1) truly empty (zero length) or 2) contain no records.  There is no distinction for some file types like FASTA that contain no metadata, but the distinction is important for file types that do.

A reader for a file type that requires metadata will rightly generate an error if metadata is not found--the file is ill-formed.  But in any format, the mere absence of records is not intrinsically an error.  It may be an error in the context of an application, but not at the library level.

Pbcore's basic readers mostly adheres to this principle.  The one exception is the cmp.h5 reader, which will throw an exception on trying to open an empty file.  The reason for that is that we don't get empty cmp.h5 files right, in our ecosystem.  When I've seen empty cmp.h5 files, they invariably are lacking important metadata tables.  Ideally this would have been fixed upstream, but 

Since the dataset concept bridges the application and data file levels, you are welcome to say that "by spec", a dataset of such and such type may not be empty, and then you can have the dataset reader throw exception on finding such a file.  But honestly then the problem is at the *writer* level---such an invalid file should never be *written*.  Adding the checks on the writer side is just good defensive programming.

**Comment by dalexander** [comment]
I will suggest one other possibility.   We need to spell out the reader API that we intend client code to use.  `openAlignmentDataset`, etc.  We can add additional API methods like `openNonemptyAlignmentDataset`, to centralize the error checking, for client code that requires nonempty.
